src/components/tl/ucp/tl_ucp_sendrecv.h

```diff
@@ -0,0 +1,115 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+#include "config.h"
+
+#ifndef UCC_TL_UCP_SENDRECV_H_
+#define UCC_TL_UCP_SENDRECV_H_
+#include "tl_ucp_tag.h"
+#include "tl_ucp_ep.h"
+#include "utils/ucc_compiler_def.h"
+void ucc_tl_ucp_send_completion_cb(void* request, ucs_status_t status,
+                                   void *user_data);
+void ucc_tl_ucp_recv_completion_cb(void* request, ucs_status_t status,
+                                   const ucp_tag_recv_info_t *info,
+                                   void *user_data);
+
+#define UCC_TL_UCP_MAKE_TAG(_tag, _rank, _id, _scope_id, _scope)       \
+    ((((uint64_t) (_tag))      << UCC_TL_UCP_TAG_BITS_OFFSET)      |   \
+     (((uint64_t) (_rank))     << UCC_TL_UCP_SENDER_BITS_OFFSET)   |   \
+     (((uint64_t) (_scope))    << UCC_TL_UCP_SCOPE_BITS_OFFSET)    |   \
+     (((uint64_t) (_scope_id)) << UCC_TL_UCP_SCOPE_ID_BITS_OFFSET) |   \
+     (((uint64_t) (_id))       << UCC_TL_UCP_ID_BITS_OFFSET))
+
+#define UCC_TL_UCP_MAKE_SEND_TAG(_tag,  _rank, _id, _scope_id, _scope) \
+    UCC_TL_UCP_MAKE_TAG(_tag, _rank, _id, _scope_id, _scope)
+
+#define UCC_TL_UCP_MAKE_RECV_TAG(_ucp_tag, _ucp_tag_mask, _tag,     \
+                                 _src, _id, _scope_id, _scope) do { \
+        ucc_assert((_tag)  <= UCC_TL_UCP_MAX_TAG);                  \
+        ucc_assert((_src)  <= UCC_TL_UCP_MAX_SENDER);               \
+        ucc_assert((_id)   <= UCC_TL_UCP_MAX_ID);                   \
+        (_ucp_tag_mask) = (uint64_t)(-1);                           \
+        (_ucp_tag) = UCC_TL_UCP_MAKE_TAG((_tag), (_src), (_id),     \
+                                         (_scope_id), (_scope));    \
+    } while(0)
+
+
+#define UCC_TL_UCP_CHECK_REQ_STATUS() do {                              \
+        if (UCS_PTR_IS_ERR(ucp_status)) {                               \
+            tl_error(UCC_TL_TEAM_LIB(team),                             \
+                     "tag %u; dest %d; team_id %u; errmsg %s",          \
+                     tag, dest_group_rank, team->id,                    \
+                     ucs_status_string(UCS_PTR_STATUS(ucp_status)));    \
+            ucp_request_cancel(UCC_TL_UCP_WORKER(team), ucp_status);    \
+            ucp_request_free(ucp_status);                               \
+            return UCC_ERR_NO_MESSAGE;                                  \
+        }                                                               \
+    } while(0)
+
+
+static inline ucc_status_t
+ucc_tl_ucp_send_nb(void *buffer, size_t msglen, ucs_memory_type_t mtype,
+                 int dest_group_rank, ucc_tl_ucp_team_t *team, uint32_t tag,
```

Comment:
Is tag argument needed? It's already there in task

Reply:
removed